File: CS/GrowthWare.WebSupport/Utilities/NameValuePairUtility.cs
using GrowthWare.Framework.BusinessData.BusinessLogicLayer;
using GrowthWare.Framework.Common;
using GrowthWare.Framework.Model.Enumerations;
using GrowthWare.Framework.Model.Profiles;
using System;
using System.Collections;
using System.Collections.ObjectModel;
using System.Data;
using System.Globalization;
using System.Web;
using System.Web.UI.WebControls;

namespace GrowthWare.WebSupport.Utilities
{
    /// <summary>
    /// Class NameValuePairUtility
    /// </summary>
    public static class NameValuePairUtility
    {
        /// <summary>
        /// Constant for the NVP Cached table
        /// </summary>
        public const string CachedNameValuePairDetailsTableName = "CachedNVPDetailsTable";

        /// <summary>
        /// Name for the cache table
        /// </summary>
        public const string CachedNameValuePairTableName = "CachedNVPTable";

        /// <summary>
        /// Deletes the detail.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns><c>true</c> if XXXX, <c>false</c> otherwise</returns>
        public static void DeleteDetail(MNameValuePairDetail profile)
        {
            if (profile == null) throw new ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!");
            MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
            BNameValuePairs mBNameValuePairs = new BNameValuePairs(mSecurityProfile);
            mBNameValuePairs.DeleteNameValuePairDetail(profile);
            CacheController.RemoveFromCache(CachedNameValuePairDetailsTableName);
        }

        /// <summary>
        /// GetNVPs will return all Name Value Pairs avalible for a given account
        /// </summary>
        /// <param name="accountId">The account the list of NVPs are for</param>
        /// <returns>Returns a data table of name value pairs for a given account</returns>
        /// <remarks></remarks>
        public static DataTable GetNameValuePairs(int accountId)
        {
            MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
            BNameValuePairs myNameValuePair = new BNameValuePairs(mSecurityProfile);
            return myNameValuePair.GetAllNameValuePair(accountId);
        }

        /// <summary>
        /// Gets the NVPID.
        /// </summary>
        /// <param name="staticName">Name of the static.</param>
        /// <returns>System.Int32.</returns>
        public static int GetNameValuePairId(string staticName)
        {
            DataView mDataView = null;
            DataTable mDataTable = null;
            int mRetValue = 0;
            try
            {
                int mDefault = -1;
                mDataTable = GetNameValuePairs(mDefault);
                mDataTable.Locale = CultureInfo.InvariantCulture;
                mDataView = mDataTable.DefaultView;
                mDataView.RowFilter = "TABLE_NAME = '" + staticName + "'";
                DataRowView mDataViewRow = mDataView[0];
                mRetValue = int.Parse(mDataViewRow["NVP_SEQ_ID"].ToString(), CultureInfo.InvariantCulture);
            }
            catch (Exception)
            {
                throw;
            }
            finally 
            {
                if (mDataTable != null) 
                {
                    mDataTable.Dispose();
                    mDataTable = null;
                }
                if (mDataView != null) 
                {
                    mDataView.Dispose();
                    mDataView = null;
                }
            }
            return mRetValue;
        }

        /// <summary>
        /// Gets the name of the NVP.
        /// </summary>
        /// <param name="nameValuePairSeqId">The NVP seq ID.</param>
        /// <returns>System.String.</returns>
        public static string GetNameValuePairName(int nameValuePairSeqId)
        {
            DataView mDataView = null;
            DataTable mDataTable = null;
            string mRetValue = string.Empty;
            try
            {
                int mDefault = -1;
                mDataTable = GetNameValuePairs(mDefault);
                mDataTable.Locale = CultureInfo.InvariantCulture;
                mDataView = mDataTable.DefaultView;
                mDataView.RowFilter = "NVP_SEQ_ID = " + nameValuePairSeqId;
                DataRowView mDataViewRow = mDataView[0];
                mRetValue = mDataViewRow["NVP_SEQ_ID"].ToString();

            }
            catch (Exception)
            {
                throw;
            }
            finally 
            {
                if (mDataTable != null) 
                {
                    mDataTable.Dispose();
                    mDataTable = null;
                }
                if (mDataView != null)
                {
                    mDataView.Dispose();
                    mDataView = null;
                }
            }
            return mRetValue;
        }

        /// <summary>
        /// GetNVPs will return all Name Value Pairs reguardless of security
        /// </summary>
        /// <param name="yourDataTable">An instance of a data table you would like populated</param>
        /// <remarks></remarks>
        public static void GetNameValuePairs(DataTable yourDataTable)
        {
            yourDataTable = (DataTable)HttpContext.Current.Cache[CachedNameValuePairTableName];
            if (yourDataTable == null)
            {
                MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
                BNameValuePairs myNameValuePair = new BNameValuePairs(mSecurityProfile);
                yourDataTable = myNameValuePair.GetAllNameValuePair();
                CacheController.AddToCacheDependency(CachedNameValuePairTableName, yourDataTable);
            }
        }

        /// <summary>
        /// Gets the NVP.
        /// </summary>
        /// <param name="nameValuePairSeqId">The NVP seq ID.</param>
        /// <returns>MNameValuePair.</returns>
        public static MNameValuePair GetNameValuePair(int nameValuePairSeqId)
        {
            MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
            BNameValuePairs myNameValuePair = new BNameValuePairs(mSecurityProfile);
            return new MNameValuePair(myNameValuePair.GetNameValuePair(nameValuePairSeqId));
        }

        /// <summary>
        /// Gets the NVP detail.
        /// </summary>
        /// <param name="nameValuePairSeqDetailId">The NVP seq det ID.</param>
        /// <param name="nameValuePairSeqId">The NVP seq ID.</param>
        /// <returns>MNameValuePairDetail.</returns>
        public static MNameValuePairDetail GetNameValuePairDetail(int nameValuePairSeqDetailId, int nameValuePairSeqId)
        {
            DataView mDataView = null;
            DataTable mDataTable = null;
            DataTable mImportTable = null;
            MNameValuePairDetail mRetVal = null;
            try
            {
                GetNameValuePairDetails(ref mDataTable, nameValuePairSeqId);
                mDataTable.Locale = CultureInfo.InvariantCulture;
                mDataView = mDataTable.DefaultView;
                mImportTable = mDataTable.Clone();
                mDataView.RowFilter = "NVP_SEQ_DET_ID = " + nameValuePairSeqDetailId;
                foreach (DataRowView drv in mDataView)
                {
                    mImportTable.ImportRow(drv.Row);
                }
                mRetVal = new MNameValuePairDetail(mImportTable.Rows[0]);
            }
            catch (Exception)
            {
                throw;
            }
            finally 
            {
                if (mDataTable != null) 
                {
                    mDataTable.Dispose();
                    mDataTable = null;
                }
                if (mDataView != null)
                {
                    mDataView.Dispose();
                    mDataView = null;
                }
                if (mImportTable != null)
                {
                    mImportTable.Dispose();
                    mImportTable = null;
                }
            }
            return mRetVal;
        }

        /// <summary>
        /// Gets the NVP details.
        /// </summary>
        /// <param name="yourDataTable">Your data table.</param>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1045:DoNotPassTypesByReference", MessageId = "0#")]
        public static void GetNameValuePairDetails(ref DataTable yourDataTable)
        {
            yourDataTable = (DataTable)HttpContext.Current.Cache[CachedNameValuePairDetailsTableName];
            if (yourDataTable == null)
            {
                MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
                BNameValuePairs myNameValuePairDetails = new BNameValuePairs(mSecurityProfile);
                yourDataTable = myNameValuePairDetails.GetAllNameValuePairDetail();
                CacheController.AddToCacheDependency(CachedNameValuePairDetailsTableName, yourDataTable);
            }
        }

        /// <summary>
        /// Gets the NVP details.
        /// </summary>
        /// <param name="yourDataTable">Your data table.</param>
        /// <param name="nameValuePairSeqId">The NVP seq ID.</param>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1045:DoNotPassTypesByReference", MessageId = "0#"), 
        System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1062:Validate arguments of public methods", MessageId = "0")]
        public static void GetNameValuePairDetails(ref DataTable yourDataTable, int nameValuePairSeqId)
        {
            DataView mDataView = null;
            DataTable mDataTable = null;
            try
            {
                GetNameValuePairDetails(ref mDataTable);
                mDataTable.Locale = CultureInfo.InvariantCulture;
                mDataTable.DefaultView.RowFilter = String.Empty;
                mDataView = new DataView(mDataTable);
                yourDataTable = mDataView.Table.Clone();
                mDataView.RowFilter = "NVP_SEQ_ID = " + nameValuePairSeqId;
                foreach (DataRowView drv in mDataView)
                {
                    yourDataTable.ImportRow(drv.Row);
                }
            }
            catch (Exception ex)
            {
                Logger mLog = Logger.Instance();
                mLog.Error(ex);
                throw;
            }
            finally 
            {
                if (mDataTable != null) 
                {
                    mDataTable.Dispose();
                    mDataTable = null;
                }
                if (mDataView != null)
                {
                    mDataView.Dispose();
                    mDataView = null;
                }
            }
        }

        /// <summary>
        /// Gets the NVP details.
        /// </summary>
        /// <param name="staticName">Name of the static.</param>
        /// <returns>DataTable.</returns>
        public static DataTable GetNameValuePairDetails(string staticName)
        {
            DataView mDataView = null;
            DataTable mDataTable = null;
            DataTable mReturnTable = null;
            try
            {
                GetNameValuePairDetails(ref mDataTable);
                mDataTable.Locale = CultureInfo.InvariantCulture;
                mDataView = mDataTable.DefaultView;
                mReturnTable = mDataView.Table.Clone();
                mDataView.RowFilter = "TABLE_NAME = '" + staticName + "'";
                foreach (DataRowView drv in mDataView)
                {
                    mReturnTable.ImportRow(drv.Row);
                }
            }
            catch (Exception)
            {
                throw;
            }
            finally 
            {
                if (mDataTable != null)
                {
                    mDataTable.Dispose();
                    mDataTable = null;
                }
                if (mDataView != null)
                {
                    mDataView.Dispose();
                    mDataView = null;
                }            
            }
            return mReturnTable;
        }

        /// <summary>
        /// Gets the NVP details.
        /// </summary>
        /// <param name="nameValuePairSeqId">The NVP seq ID.</param>
        /// <returns>DataTable.</returns>
        public static DataTable GetNameValuePairDetails(int nameValuePairSeqId)
        {
            DataView mDataView = null;
            DataTable mDataTable = null;
            DataTable mReturnTable = null;
            try
            {
                GetNameValuePairDetails(ref mDataTable);
                mDataTable.Locale = CultureInfo.InvariantCulture;
                mDataView = mDataTable.DefaultView;
                mReturnTable = mDataView.Table.Clone();
                mDataView.RowFilter = "NVP_SEQ_ID = " + nameValuePairSeqId;
                foreach (DataRowView drv in mDataView)
                {
                    mReturnTable.ImportRow(drv.Row);
                }
            }
            catch (Exception)
            {
                throw;
            }
            finally 
            {
                if (mDataTable != null)
                {
                    mDataTable.Dispose();
                    mDataTable = null;
                }
                if (mDataView != null)
                {
                    mDataView.Dispose();
                    mDataView = null;
                }                 
            }
            return mReturnTable;
        }

        /// <summary>
        /// Retrieves the selected roles for the name value pair
        /// </summary>
        /// <param name="nameValuePairSeqId">The primary key of the name value pair</param>
        /// <returns>String array list of roles</returns>
        /// <remarks></remarks>
        public static string[] GetSelectedRoles(int nameValuePairSeqId)
        {
            MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
            BNameValuePairs myNameValuePair = new BNameValuePairs(mSecurityProfile);
            DataTable myDataTable = myNameValuePair.GetNameValuePairRoles(nameValuePairSeqId);
            return GetStringArrayList(myDataTable, "Roles");
        }

        /// <summary>
        /// Retrieves the selected groups for the name value pair
        /// </summary>
        /// <param name="nameValuePairSeqId">The primary key of the name value pair</param>
        /// <returns>String array list of groups</returns>
        /// <remarks></remarks>
        public static string[] GetSelectedGroups(int nameValuePairSeqId)
        {
            MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
            BNameValuePairs myNameValuePair = new BNameValuePairs(mSecurityProfile);
            DataTable myDataTable = myNameValuePair.GetNameValuePairGroups(nameValuePairSeqId);
            return GetStringArrayList(myDataTable, "Groups");
        }

        /// <summary>
        /// Add all the roles or groups to a string array list to bind to the List picker control.
        /// </summary>
        /// <param name="yourDataTable">The DataTable containing the data to bind</param>
        /// <param name="rowName">The name of the row to add to the array list</param>
        /// <returns>String array list</returns>
        /// <remarks>See the GrowthWare.CustomWebControls.ListPicker object</remarks>
        private static string[] GetStringArrayList(DataTable yourDataTable, string rowName)
        {
            ArrayList mRetrunArrayList = new ArrayList();
            DataRow myDR = null;
            foreach (DataRow myDR_loopVariable in yourDataTable.Rows)
            {
                myDR = myDR_loopVariable;
                mRetrunArrayList.Add(myDR[rowName].ToString());
            }
            return (string[])mRetrunArrayList.ToArray(typeof(string));
        }

        /// <summary>
        /// Inserts or updates a name value pair given a NameValuePair profile
        /// </summary>
        /// <param name="profile">Details of the profile to added or updated</param>
        /// <returns>The Primary sequence ID</returns>
        /// <remarks></remarks>
        public static int Save(MNameValuePair profile)
        {
            if (profile == null) throw new ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!");
            int mRetVal = -1;
            MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
            BNameValuePairs myNameValuePair = new BNameValuePairs(mSecurityProfile);
            CacheController.RemoveFromCache(CachedNameValuePairTableName);
            CacheController.RemoveFromCache(CachedNameValuePairDetailsTableName);
            mRetVal = myNameValuePair.Save(profile);
            return mRetVal;
        }

        /// <summary>
        /// Saves the detail.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns>System.Int32.</returns>
        public static void SaveDetail(MNameValuePairDetail profile)
        {
            if (profile == null) throw new ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!");
            MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
            BNameValuePairs mBNameValuePairs = new BNameValuePairs(mSecurityProfile);
            CacheController.RemoveFromCache(CachedNameValuePairDetailsTableName);
            mBNameValuePairs.SaveNameValuePairDetail(profile);
        }

        /// <summary>
        /// Searches the specified search criteria.
        /// </summary>
        /// <param name="searchCriteria"></param>
        /// <returns></returns>
        public static DataTable Search(MSearchCriteria searchCriteria)
        {
            if (searchCriteria == null) throw new ArgumentNullException("searchCriteria", "searchCriteria cannot be a null reference (Nothing in Visual Basic)!");
            MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
            BNameValuePairs mBNameValuePairs = new BNameValuePairs(mSecurityProfile);
            return mBNameValuePairs.Search(searchCriteria);
        }

        /// <summary>
        /// SetDropSelection will set the selected index given the drop down list and the desired value
        /// </summary>
        /// <param name="theDropDown">The drop down list to set</param>
        /// <param name="selectedValue">The value to set the drop box to</param>
        /// <remarks></remarks>
        public static void SetDropSelection(ListControl theDropDown, string selectedValue)
        {
            if (theDropDown == null) throw new ArgumentNullException("theDropDown", "theDropDown cannot be a null reference (Nothing in Visual Basic)!");
            if (string.IsNullOrEmpty(selectedValue)) throw new ArgumentNullException("selectedValue", "selectedValue cannot be a null reference (Nothing in Visual Basic)!");
            try
            {
                int X = 0;
                for (X = 0; X <= theDropDown.Items.Count - 1; X++)
                {
                    if (theDropDown.Items[X].Value == selectedValue)
                    {
                        theDropDown.SelectedIndex = X;
                        break; // TODO: might not be correct. Was : Exit For
                    }
                }
            }
            catch (Exception)
            {
                throw;
            }
        }

        /// <summary>
        /// Updates the groups.
        /// </summary>
        /// <param name="nameValuePairId">The NV p_ ID.</param>
        /// <param name="securityEntityId">The security entity ID.</param>
        /// <param name="commaSeparatedGroups">The comma separated groups.</param>
        /// <param name="nameValuePairProfile">MNameValuePair.</param>
        public static void UpdateGroups(int nameValuePairId, int securityEntityId, string commaSeparatedGroups, MNameValuePair nameValuePairProfile)
        {
            if (nameValuePairProfile == null) throw new ArgumentNullException("nameValuePairProfile", "nameValuePairProfile cannot be a null reference (Nothing in Visual Basic)!");
            MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
            BNameValuePairs mNameValuePair = new BNameValuePairs(mSecurityProfile);
            mNameValuePair.UpdateGroups(nameValuePairId, securityEntityId, commaSeparatedGroups, nameValuePairProfile);
        }

        /// <summary>
        /// Updates the roles.
        /// </summary>
        /// <param name="nameValuePairId">The NV p_ ID.</param>
        /// <param name="securityEntityId">The security entity ID.</param>
        /// <param name="commaSeparatedRoles">The comma separated roles.</param>
        /// <param name="nameValuePairProfile">MNameValuePair.</param>
        public static void UpdateRoles(int nameValuePairId, int securityEntityId, string commaSeparatedRoles, MNameValuePair nameValuePairProfile)
        {
            if (nameValuePairProfile == null) throw new ArgumentNullException("nameValuePairProfile", "nameValuePairProfile cannot be a null reference (Nothing in Visual Basic)!");
            MSecurityEntityProfile mSecurityProfile = SecurityEntityUtility.CurrentProfile();
            BNameValuePairs myNameValuePair = new BNameValuePairs(mSecurityProfile);
            myNameValuePair.UpdateRoles(nameValuePairId, securityEntityId, commaSeparatedRoles, nameValuePairProfile);
        }

    }
}

File: VB/GrowthWare.WebSupport/Utilities/NameValuePairUtility.vb
Imports GrowthWare.Framework.BusinessData.BusinessLogicLayer
Imports GrowthWare.Framework.Common
Imports GrowthWare.Framework.Model.Profiles
Imports System.Web
Imports System.Web.UI
Imports System.Globalization
Imports System.Web.UI.WebControls

Namespace Utilities
    ''' <summary>
    ''' Class NameValuePairUtility
    ''' </summary>
    Public Module NameValuePairUtility
        ''' <summary>
        ''' Cached Name Value Pair Details Table Name
        ''' </summary>
        Public Const CachedNameValuePairDetailsTableName As String = "CachedNVPDetailsTable"

        ''' <summary>
        ''' Cached Name Value Pair Table Name
        ''' </summary>
        Public Const CachedNameValuePairTableName As String = "CachedNVPTable"

        ''' <summary>
        ''' Deletes the detail.
        ''' </summary>
        ''' <param name="profile">The profile.</param>
        Public Sub DeleteDetail(ByVal profile As MNameValuePairDetail)
            If profile Is Nothing Then Throw New ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!")
            Dim mBNameValuePairs As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
            mBNameValuePairs.DeleteNameValuePairDetail(profile)
            CacheController.RemoveFromCache(CachedNameValuePairDetailsTableName)
        End Sub

        ''' <summary>
        ''' GetNVPs will return all Name Value Pairs avalible for a given account
        ''' </summary>
        ''' <param name="accountId">The account ID.</param>
        ''' <returns>Returns a data table of name value pairs for a given account</returns>
        Public Function AllNameValuePairs(ByVal accountId As Integer) As DataTable
            Dim mNameValuePair As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
            Return mNameValuePair.AllNameValuePairs(accountId)
        End Function

        ''' <summary>
        ''' Gets the NVPID.
        ''' </summary>
        ''' <param name="staticName">Name of the static.</param>
        ''' <returns>System.Int32.</returns>
        Public Function GetNameValuePairId(ByVal staticName As String) As Integer
            Dim mDataView As DataView = Nothing
            Dim mDataTable As DataTable = Nothing
            Dim mRetValue As Integer = 0
            Try
                mDataTable = AllNameValuePairs(-1)
                mDataTable.Locale = CultureInfo.InvariantCulture
                mDataView = mDataTable.DefaultView
                mDataView.RowFilter = "TABLE_NAME = '" & staticName & "'"
                Dim mDataViewRow As DataRowView = mDataView.Item(0)
                mRetValue = Integer.Parse(mDataViewRow.Item("NVP_SEQ_ID").ToString(), CultureInfo.InvariantCulture)
            Catch ex As Exception
                Throw
            Finally
                If Not mDataTable Is Nothing Then
                    mDataTable.Dispose()
                    mDataTable = Nothing
                End If
                If Not mDataView Is Nothing Then
                    mDataView.Dispose()
                    mDataView = Nothing
                End If
            End Try
            Return mRetValue
        End Function

        ''' <summary>
        ''' Gets the name of the NVP.
        ''' </summary>
        ''' <param name="nameValuePairSeqId">The NVP seq ID.</param>
        ''' <returns>System.String.</returns>
        Public Function GetNameValuePairName(ByVal nameValuePairSeqId As Integer) As String
            Dim mDataView As DataView = Nothing
            Dim mDataTable As DataTable = Nothing
            Dim mRetValue As String = String.Empty
            Try
                mDataTable = AllNameValuePairs(-1)
                mDataTable.Locale = CultureInfo.InvariantCulture
                mDataView = mDataTable.DefaultView
                mDataView.RowFilter = "NVP_SEQ_ID = " & nameValuePairSeqId
                Dim mDataViewRow As DataRowView = mDataView.Item(0)
                mRetValue = mDataViewRow.Item("NVP_SEQ_ID").ToString()
            Catch ex As Exception
                Throw
            Finally
                If Not mDataTable Is Nothing Then
                    mDataTable.Dispose()
                    mDataTable = Nothing
                End If
                If Not mDataView Is Nothing Then
                    mDataView.Dispose()
                    mDataView = Nothing
                End If
            End Try
            Return mRetValue
        End Function

        ''' <summary>
        ''' GetNVPs will return all Name Value Pairs reguardless of security
        ''' </summary>
        ''' <param name="yourDataTable">An instance of a data table you would like populated</param>
        Public Sub GetNameValuePairs(ByVal yourDataTable As DataTable)
            yourDataTable = CType(HttpContext.Current.Cache(CachedNameValuePairTableName), DataTable)
            If yourDataTable Is Nothing Then
                Dim mNameValuePair As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
                yourDataTable = mNameValuePair.AllNameValuePairs()
                CacheController.AddToCacheDependency(CachedNameValuePairTableName, yourDataTable)
            End If
        End Sub

        ''' <summary>
        ''' Gets the NVP.
        ''' </summary>
        ''' <param name="nameValuePairSeqId">The NVP seq ID.</param>
        ''' <returns>MNameValuePair.</returns>
        Public Function GetNameValuePair(ByVal nameValuePairSeqId As Integer) As MNameValuePair
            Dim mNameValuePair As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
            Return New MNameValuePair(mNameValuePair.GetNameValuePair(nameValuePairSeqId))
        End Function

        ''' <summary>
        ''' Gets the NVP detail.
        ''' </summary>
        ''' <param name="nameValuePairDetailSeqId">The NVP seq det ID.</param>
        ''' <param name="nameValuePairSeqId">The NVP seq ID.</param>
        ''' <returns>MNameValuePairDetail.</returns>
        Public Function GetNameValuePairDetail(ByVal nameValuePairDetailSeqId As Integer, ByVal nameValuePairSeqId As Integer) As MNameValuePairDetail
            Dim mDataView As DataView = Nothing
            Dim mDataTable As DataTable = Nothing
            Dim mImportTable As DataTable = Nothing
            Try
                GetNameValuePairDetails(mDataTable, nameValuePairSeqId)
                mDataTable.Locale = CultureInfo.InvariantCulture
                mDataView = mDataTable.DefaultView
                mImportTable = mDataTable.Clone
                mDataView.RowFilter = "NVP_SEQ_DET_ID = " & nameValuePairDetailSeqId
                For Each drv As DataRowView In mDataView
                    mImportTable.ImportRow(drv.Row)
                Next
            Catch ex As Exception
                Throw
            Finally
                If Not mDataTable Is Nothing Then
                    mDataTable.Dispose()
                    mDataTable = Nothing
                End If
                If Not mDataView Is Nothing Then
                    mDataView.Dispose()
                    mDataView = Nothing
                End If
            End Try
            Return New MNameValuePairDetail(mImportTable.Rows(0))
        End Function

        ''' <summary>
        ''' Gets the NVP details.
        ''' </summary>
        ''' <param name="yourDataTable">Your data table.</param>
        <System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1045:DoNotPassTypesByReference", MessageId:="0#")>
        Public Sub GetNameValuePairDetails(ByRef yourDataTable As DataTable)
            yourDataTable = CType(HttpContext.Current.Cache(CachedNameValuePairDetailsTableName), DataTable)
            If yourDataTable Is Nothing Then
                Dim mNameValuePairDetails As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
                yourDataTable = mNameValuePairDetails.GetAllNameValuePairDetail()
                CacheController.AddToCacheDependency(CachedNameValuePairDetailsTableName, yourDataTable)
            End If
        End Sub

        ''' <summary>
        ''' Gets the NVP details.
        ''' </summary>
        ''' <param name="yourDataTable">Your data table.</param>
        ''' <param name="nameValuePairSeqId">The NVP seq ID.</param>
        <System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1045:DoNotPassTypesByReference", MessageId:="0#")>
        <System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1062:Validate arguments of public methods", MessageId:="0")>
        Public Sub GetNameValuePairDetails(ByRef yourDataTable As DataTable, ByVal nameValuePairSeqId As Integer)
            Dim mDataView As DataView = Nothing
            Dim mDataTable As DataTable = Nothing
            Try
                GetNameValuePairDetails(mDataTable)
                mDataTable.Locale = CultureInfo.InvariantCulture
                mDataTable.DefaultView.RowFilter = String.Empty
                mDataView = New DataView(mDataTable)
                yourDataTable = mDataView.Table.Clone()
                mDataView.RowFilter = "NVP_SEQ_ID = " & nameValuePairSeqId
                For Each drv As DataRowView In mDataView
                    yourDataTable.ImportRow(drv.Row)
                Next
            Catch ex As Exception
                Dim mLog As Logger = Logger.Instance()
                mLog.Error(ex)
                Throw
            Finally
                If Not mDataTable Is Nothing Then
                    mDataTable.Dispose()
                    mDataTable = Nothing
                End If
                If Not mDataView Is Nothing Then
                    mDataView.Dispose()
                    mDataView = Nothing
                End If
            End Try

        End Sub

        ''' <summary>
        ''' Gets the NVP details.
        ''' </summary>
        ''' <param name="staticName">Name of the static.</param>
        ''' <returns>DataTable.</returns>
        Public Function GetNameValuePairDetails(ByVal staticName As String) As DataTable
            Dim mDataView As DataView = Nothing
            Dim mDataTable As DataTable = Nothing
            Dim mReturnTable As DataTable = Nothing
            Try
                GetNameValuePairDetails(mDataTable)
                mDataTable.Locale = CultureInfo.InvariantCulture
                mDataView = mDataTable.DefaultView
                mReturnTable = mDataView.Table.Clone()
                mDataView.RowFilter = "TABLE_NAME = '" & staticName & "'"
                For Each drv As DataRowView In mDataView
                    mReturnTable.ImportRow(drv.Row)
                Next
            Catch ex As Exception
                Throw
            Finally
                If Not mDataTable Is Nothing Then
                    mDataTable.Dispose()
                    mDataTable = Nothing
                End If
                If Not mDataView Is Nothing Then
                    mDataView.Dispose()
                    mDataView = Nothing
                End If
            End Try
            Return mReturnTable
        End Function

        ''' <summary>
        ''' Gets the NVP details.
        ''' </summary>
        ''' <param name="nameValuePairSeqId">The NVP seq ID.</param>
        ''' <returns>DataTable.</returns>
        Public Function GetNameValuePairDetails(ByVal nameValuePairSeqId As Integer) As DataTable
            Dim mDataView As DataView = Nothing
            Dim mDataTable As DataTable = Nothing
            Dim mReturnTable As DataTable = Nothing
            Try
                GetNameValuePairDetails(mDataTable)
                mDataTable.Locale = CultureInfo.InvariantCulture
                mDataView = mDataTable.DefaultView
                mReturnTable = mDataView.Table.Clone()
                mDataView.RowFilter = "NVP_SEQ_ID = " & nameValuePairSeqId
                For Each drv As DataRowView In mDataView
                    mReturnTable.ImportRow(drv.Row)
                Next

            Catch ex As Exception
                Throw
            Finally
                If Not mDataTable Is Nothing Then
                    mDataTable.Dispose()
                    mDataTable = Nothing
                End If
                If Not mDataView Is Nothing Then
                    mDataView.Dispose()
                    mDataView = Nothing
                End If
            End Try
            Return mReturnTable
        End Function

        ''' <summary>
        ''' Gets the selected roles.
        ''' </summary>
        ''' <param name="nameValuePairSeqId">The name value pair seq ID.</param>
        ''' <returns>System.String[][].</returns>
        Public Function GetSelectedRoles(ByVal nameValuePairSeqId As Integer) As String()
            Dim mNameValuePair As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
            Dim mDataTable As DataTable = mNameValuePair.GetNameValuePairRoles(nameValuePairSeqId)
            Return GetStringArrayList(mDataTable, "Roles")
        End Function

        ''' <summary>
        ''' Gets the selected groups.
        ''' </summary>
        ''' <param name="nameValuePairSeqId">The name value pair seq ID.</param>
        ''' <returns>System.String[][].</returns>
        Public Function GetSelectedGroups(ByVal nameValuePairSeqId As Integer) As String()
            Dim mNameValuePair As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
            Dim mDataTable As DataTable = mNameValuePair.GetNameValuePairGroups(nameValuePairSeqId)
            Return GetStringArrayList(mDataTable, "Groups")
        End Function

        ''' <summary>
        ''' Gets the string array list.
        ''' </summary>
        ''' <param name="yourDataTable">Your DT.</param>
        ''' <param name="rowName">Name of the row.</param>
        ''' <returns>System.String[][].</returns>
        Private Function GetStringArrayList(ByVal yourDataTable As DataTable, ByVal rowName As String) As String()
            Dim mRetrunArrayList As New ArrayList
            Dim mDR As DataRow
            For Each mDR In yourDataTable.Rows
                mRetrunArrayList.Add(mDR(rowName).ToString)
            Next
            Return CType(mRetrunArrayList.ToArray(GetType(String)), String())
        End Function

        ''' <summary>
        ''' Saves the specified profile.
        ''' </summary>
        ''' <param name="profile">The profile.</param>
        Public Function Save(ByVal profile As MNameValuePair) As Integer
            If profile Is Nothing Then Throw New ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!")
            Dim mNameValuePair As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
            Dim mRetVal As Integer = -1
            mRetVal = mNameValuePair.Save(profile)
            CacheController.RemoveFromCache(CachedNameValuePairTableName)
            CacheController.RemoveFromCache(CachedNameValuePairDetailsTableName)
            Return mRetVal
        End Function

        ''' <summary>
        ''' Saves the detail.
        ''' </summary>
        ''' <param name="profile">The profile.</param>
        Public Sub SaveDetail(ByVal profile As MNameValuePairDetail)
            If profile Is Nothing Then Throw New ArgumentNullException("profile", "profile cannot be a null reference (Nothing in Visual Basic)!")
            Dim mBNameValuePairs As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
            mBNameValuePairs.SaveNameValuePairDetail(profile)
            CacheController.RemoveFromCache(CachedNameValuePairDetailsTableName)
        End Sub

        ''' <summary>
        ''' Searches the specified search criteria.
        ''' </summary>
        ''' <param name="searchCriteria">The search criteria.</param>
        ''' <returns>DataTable.</returns>
        Public Function Search(ByVal searchCriteria As MSearchCriteria) As DataTable
            If searchCriteria Is Nothing Then Throw New ArgumentNullException("searchCriteria", "searchCriteria cannot be a null reference (Nothing in Visual Basic)!")
            Try
                Dim mBNameValuePairs As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
                Return mBNameValuePairs.Search(searchCriteria)
            Catch ex As IndexOutOfRangeException
                'no data is not a problem
                Return Nothing
            End Try
        End Function

        ''' <summary>
        ''' Sets the drop selection.
        ''' </summary>
        ''' <param name="theDropDown">The drop down.</param>
        ''' <param name="selectedVale">The selected vale.</param>
        Public Sub SetDropSelection(ByVal theDropDown As ListControl, ByVal selectedVale As String)
            If theDropDown Is Nothing Then Throw New ArgumentNullException("theDropDown", "theDropDown cannot be a null reference (Nothing in Visual Basic)!")
            If String.IsNullOrEmpty(selectedVale) Then Throw New ArgumentNullException("selectedVale", "selectedVale cannot be a null reference (Nothing in Visual Basic)!")
            Try
                Dim X As Integer
                For X = 0 To theDropDown.Items.Count - 1
                    If theDropDown.Items(X).Value = selectedVale Then
                        theDropDown.SelectedIndex = X
                        Exit For
                    End If
                Next
            Catch ex As Exception
                Throw
            End Try
        End Sub

        ''' <summary>
        ''' Updates the groups.
        ''' </summary>
        ''' <param name="nameValuePairId">The name value pair Id.</param>
        ''' <param name="securityEntityId">The security entity Id.</param>
        ''' <param name="commaSeparatedGroups">The comma Separated groups.</param>
        ''' <param name="nameValuePairProfile">MNameValuePair</param>
        Public Sub UpdateGroups(ByVal nameValuePairId As Integer, ByVal securityEntityId As Integer, ByVal commaSeparatedGroups As String, ByVal nameValuePairProfile As MNameValuePair)
            Dim mNameValuePair As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
            mNameValuePair.UpdateGroups(nameValuePairId, securityEntityId, commaSeparatedGroups, nameValuePairProfile)
        End Sub

        ''' <summary>
        ''' Updates the roles.
        ''' </summary>
        ''' <param name="nameValuePairId">The NV p_ ID.</param>
        ''' <param name="securityEntityId">The security entity ID.</param>
        ''' <param name="commaSeparatedRoles">The comma Separated roles.</param>
        ''' <param name="nameValuePairProfile">nameValuePairProfile</param>
        Public Sub UpdateRoles(ByVal nameValuePairId As Integer, ByVal securityEntityId As Integer, ByVal commaSeparatedRoles As String, ByVal nameValuePairProfile As MNameValuePair)
            Dim mNameValuePair As New BNameValuePairs(SecurityEntityUtility.CurrentProfile, ConfigSettings.CentralManagement)
            mNameValuePair.UpdateRoles(nameValuePairId, securityEntityId, commaSeparatedRoles, nameValuePairProfile)
        End Sub
    End Module
End Namespace
